### PR TITLE
Hint support

### PR DESCRIPTION
**Pull request by AntyMew**
Supports writing hints to navigation room messages.

The text encoding code isn't the cleanest, a more flexible solution will probably be needed for when we start supporting more custom strings

### CODE REVIEW

File: src/mfr_patcher/__main__.py [hunk 1]
```diff
@@ -50,6 +51,16 @@
     item_patcher = ItemPatcher(rom, loc_settings)
     item_patcher.write_items()
 
+    # get hints
+    hints = None
+    if "Hints" in patch_data:
+        hints = Hints.from_json(patch_data["Hints"])
```
Comment (Miepee) [inline src/mfr_patcher/__main__.py:57]: You can shorten this to `hints = Hints.from_json(patch_data.get("Hints", None)`
The second get parameter is the default variable for when the key does not exist

That way you could also inline it with the check below. Up to you if you wanna do that

File: src/mfr_patcher/rom.py [hunk 2]
```diff
@@ -392,3 +392,19 @@ def file_screen_text(self) -> int:
...
+            elif self.region == Region.J:
+                return 0x79904
+            # TODO: find address in PAL
+
+    def navigation_text(self) -> int:
+        """Returns the address of the navigation room text pointers."""
+        if self.game == Game.MF:
+            if self.region == Region.U:
+                return 0x79C0F0
+            # TODO: find addresses in NTSC-J and PAL
```
Comment (Miepee) [inline src/mfr_patcher/rom.py:410]: `TODO: Find addresses for ZM`? Dunno whether they make sense here, but the other functions do seem to have one.

File: src/mfr_patcher/__main__.py (hunk 1 above)
Comment (AntyMew) [reply src/mfr_patcher/__main__.py:57]: The `from_json` method doesn't accept `None` as a valid input. I'm not entirely sure about doing it this way either, but it's consistent with how the lines above handle things

Comment (Miepee) [reply src/mfr_patcher/__main__.py:57]: Ah ok

File: src/mfr_patcher/rom.py (hunk 2 above)
Comment (Miepee) [reply src/mfr_patcher/rom.py]: might throw a value error then?

File: src/mfr_patcher/data/schema.json [hunk 3]
```diff
@@ -140,6 +140,35 @@
         },
         "SkipDoorTransitions": {
             "type": "boolean"
+        },
+        "Hints": {
+            "type": "array",
```
Comment (Miepee) [inline src/mfr_patcher/data/schema.json:145]: Can you add a `description`, `uniqueItems` and `default` field here?

File: src/mfr_patcher/data/schema.json [hunk 4]
```diff
@@ -140,6 +140,35 @@
...
+            "items": {
+                "type": "object",
```
Comment (Miepee) [inline src/mfr_patcher/data/schema.json:147]: A `description` field could be added here. could also be ommitted, up to you.

File: src/mfr_patcher/data/schema.json [hunk 5]
```diff
@@ -140,6 +140,35 @@
...
+                "properties": {
+                    "Language": {
+                        "$ref": "#/$defs/ValidLanguages"
+                    },
+                    "Text": {
+                        "type": "object",
```
Comment (Miepee) [inline src/mfr_patcher/data/schema.json:153]: Can you add a `description` field here?

File: src/mfr_patcher/data/schema.json [hunk 6]
```diff
@@ -140,6 +140,35 @@
...
+                            "MainDeckEast": { "type": "string" },
+                            "OperationsDeck": { "type": "string" },
+                            "Sector1Entrance": { "type": "string" },
+                            "Sector2Entrance": { "type": "string" },
+                            "Sector3Entrance": { "type": "string" },
+                            "Sector4Entrance": { "type": "string" },
+                            "Sector5Entrance": { "type": "string" },
+                            "Sector6Entrance": { "type": "string" },
+                            "AuxiliaryPower": { "type": "string" },
+                            "RestrictedLabs": { "type": "string" }
```
Comment (Miepee) [inline src/mfr_patcher/data/schema.json:165]: Do you want these to be required? Would be a bit weird to have:
```json
"Hints": [
 {
  "Text": {
    "RestrictedLabs": "only entry"
  },
  "Language": "English"
  }
]
```
, no?

File: src/mfr_patcher/data/schema.json [hunk 7]
```diff
@@ -238,6 +267,18 @@
                 "ScrewAttack",
                 "IceBeam"
             ]
+        },
+        "ValidLanguages": {
+            "type": "string",
```
Comment (Miepee) [inline src/mfr_patcher/data/schema.json:272]: Can you add a `description` field here?

**Review by Miepee — COMMENTED** [review]
Some more comments regarding schema. Feel free to look at the other PR or ping me if these are unclear
